### PR TITLE
add unittest.case._AssertRaisesBaseContext class

### PR DESCRIPTION
**Pull request by tungol**
related to https://github.com/python/typeshed/issues/3968

### CODE REVIEW

File: stdlib/unittest/case.pyi [hunk 1]
```diff
@@ -27,6 +27,23 @@ DIFF_OMITTED: str
...
+    def __init__(
+        self,
+        expected: type[BaseException] | tuple[type[BaseException], ...],
+        test_case: TestCase,
+        expected_regex: str | Pattern[str] | None = None,
+    ) -> None: ...
+
+    # This returns Self if args is the empty list, and None otherwise.
+    # but it's not possible to construct an overload which expresses that
+    def handle(self, name: str, args: list[Any], kwargs: dict[str, Any]) -> Any: ...
```
Comment (tungol) [inline stdlib/unittest/case.pyi:45]: I'm about 90% sure this comment is accurate, but if I'm wrong and there's a way to do this overload, I'd be glad to hear it.

File: stdlib/unittest/_log.pyi [hunk 2]
```diff
@@ -2,15 +2,15 @@ import logging
...
+from unittest.case import TestCase, _AssertRaisesBaseContext
 
 _L = TypeVar("_L", None, _LoggingWatcher)
 
 class _LoggingWatcher(NamedTuple):
     records: list[logging.LogRecord]
     output: list[str]
 
-class _AssertLogsContext(Generic[_L]):
+class _AssertLogsContext(_AssertRaisesBaseContext, Generic[_L]):
```
Comment (srittau) [inline stdlib/unittest/_log.pyi:13]: In Python 3.11 and 3.12 at least, this derives directly from `_BaseTestCaseContext`, not `_AssertRaisesBaseContext`.

File: stdlib/unittest/case.pyi [hunk 3]
```diff
@@ -27,6 +27,23 @@ DIFF_OMITTED: str
 class _BaseTestCaseContext:
     def __init__(self, test_case: TestCase) -> None: ...
```
Comment (srittau) [inline stdlib/unittest/case.pyi:28]:
```suggestion
class _BaseTestCaseContext:
    test_case: TestCase
    def __init__(self, test_case: TestCase) -> None: ...
```

File: stdlib/unittest/case.pyi [hunk 4]
```diff
@@ -27,6 +27,23 @@ DIFF_OMITTED: str
...
 
+class _AssertRaisesBaseContext(_BaseTestCaseContext):
+    expected: type[BaseException] | tuple[type[BaseException], ...]
+    expected_regex: str | Pattern[str] | None
```
Comment (srittau) [inline stdlib/unittest/case.pyi:32]: This is compiled in `__init__`:

```suggestion
    expected_regex: Pattern[str] | None
```

**Review by srittau — CHANGES_REQUESTED** [review]
Thanks, comments below.

File: stdlib/unittest/case.pyi (hunk 3 above)
Comment (tungol) [reply stdlib/unittest/case.pyi:28]: good catch - I can remove `test_case: TestCase` from some of the child classes with that too.

File: stdlib/unittest/_log.pyi (hunk 2 above)
Comment (tungol) [reply stdlib/unittest/_log.pyi]: It's that way on all versions - best I can figure is a copy/paste error of some kind. Fixed now, at any rate.

**Review by srittau — APPROVED** [review]

**Pull request merged**